### ISSUE TITLE
Data leakage in the validation step?

### ISSUE THREAD

**Issue by Sjyhne** [woodfrog/heat#14]
Hi,

I might be mistaken, but I suspect there is a potential data leakage in the validation/evaluation step in the training script. During training you are using the ground truth labels for label embeddings, which again is used in the edge-prediction part before each of the hybrid and geometric decoders. 

In the edge_models.py file at 234-237.
```
# Generate masked training info
if not do_inference:
    filtered_gt_values = self.generate_gt_masking(filtered_labels, filtered_mask)
else:
    filtered_gt_values = filtered_labels  # Using ground truth during inference!
```

So, even if you set the `do_inference = True` flag, it will still be using the actual ground truth information from the labels. 

I can see that the infer.py file does not do it like this, but rather does the prediction in an iterative manner where you use the most confident previous predictions as the `gt_info`, so that part seems fine. 

Just trying to figure out if I am mistaken and have missed some part of the flow that eliminates this leakage, or if it should be fixed.

**Comment by woodfrog** [comment]
The validation step simply follows the masked training setting and only checks the accuracy of the unknown (i.e., masked) edges (Sec.4.4 of the paper). Those known edges (i.e., not masked) are leaked by design. 

The ```infer.py```  is what we used to do the final quantitative/qualitative evaluation.

**Comment by Sjyhne** [comment]
Yeah, I was kinda expecting you to use the infer for the qualitative/quantitative results, so that is great.

I was just expecting you would use the same inference method in the validation step as you do in the `infer.py` file, as that would be more "representative" of what the performance will be when not having access to the gt.

Thanks for the response!

**Comment by woodfrog** [comment]
Yeah that makes sense to me.

I remember I didn't use the validation results to select checkpoints. Those validation results were only used for monitoring the training progress, so either using the masked training setting or the actual test setting (no gt labels in the input) works for this purpose.
